pystac/stac_object.py

```diff
@@ -136,6 +137,28 @@ def remove_hierarchical_links(self, add_canonical: bool = False) -> List[Link]:
         self.links = keep
         return remove
 
+    def target_in_hierarchy(self, target: Union[str, STACObject]) -> bool:
+        """Recursively collects all the targets referred to by the hierarchical
+        links of the current STACObject.
+
+        Returns:
+            Set[Union[str, STACObject]]: All encountered targets
+        """
+
+        def traverse(
+            obj: Union[str, STACObject], visited: Set[Union[str, STACObject]]
+        ) -> Set[Union[str, STACObject]]:
+            if isinstance(obj, str):
+                return visited
+
+            hierarchical_links = [link for link in obj.links if link.is_hierarchical()]
+            new_targets = set([link.target for link in hierarchical_links]) - visited
+            for target in new_targets:
+                visited = traverse(target, visited.union(set([target])))
+            return visited
+
+        return target in traverse(self, set([self]))
```

Comment:
This should be refactored to short-circuit when the target is encountered, so it doesn't walk the whole tree if it doesn't need to.

Reply:
🤦